### PR TITLE
Most viewed endpoint - correct logic

### PR DESCRIPTION
**Pull request by GHaberis**
## What does this change?

Copies the logic from [frontend](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/onward/popular.js) for deciding the endpoint to hit for "most read" article list.

These endpoints can return an array of lists, so this PR also handles that and if there are multiple lists builds tabs for toggling between the 2 lists, I've made an assumption in the styles there will never be more than 2 lists, I've never seen this on frontend.

**Two lists:**

mostviewed

**One list:**

screen shot 2018-10-10 at 10 48 52

### CODE REVIEW

File: frontend/components/MostViewed.tsx [hunk 1]
```diff
@@ -147,49 +148,199 @@ const headlineLink = css`
...
+        this.setState({
+            selectedTabIndex: index,
+        });
+    }
+
+    public render() {
+        return (
+            <div className={container}>
+                <h2 className={heading}>Most Viewed</h2>
+                <AsyncClientComponent f={this.fetchTrails}>
```
Comment (SiAdcock) [inline frontend/components/MostViewed.tsx]: Is `this` bound correctly?

Comment (GHaberis) [reply frontend/components/MostViewed.tsx]: As described in this [post](https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript#use-instance-functions) from Microsoft's typescript wiki you can correctly keep the context of `this` by using an 'instance arrow function" to define a class member, as I've done here with `fetchTrails`.

File: frontend/components/MostViewed.tsx [hunk 2]
```diff
@@ -147,49 +148,199 @@ const headlineLink = css`
...
+                                                        Most viewed{' '}
+                                                    </span>
+                                                    {tab.heading}
+                                                </button>
+                                            </li>
+                                        ))}
+                                    </ol>
+                                )}
+                            {(data || []).map((tab, i) => (
+                                <ul
```
Comment (SiAdcock) [inline frontend/components/MostViewed.tsx]: Ought this be an ordered list?

File: frontend/components/MostViewed.tsx [hunk 3]
```diff
@@ -147,49 +148,199 @@ const headlineLink = css`
...
+
+    public fetchTrails: () => Promise<Tab[]> = () => {
+        const sectionsWithoutPopular = ['info', 'global'];
+        const hasSection =
+            this.props.sectionName &&
+            !sectionsWithoutPopular.includes(this.props.sectionName);
+        const endpoint = `/most-read${
+            hasSection ? `/${this.props.sectionName}` : ''
+        }.json`;
+        return new Promise((resolve, reject) => {
```
Comment (SiAdcock) [inline frontend/components/MostViewed.tsx]: A philosophical question. Does this logic belong here, at the presentational layer, or further up at the data layer in `parse-capi`? I'm not suggesting we move it, but we ought to start thinking about how much logic is executed down here among the presentational code

Comment (GHaberis) [reply frontend/components/MostViewed.tsx]: If it were to exist in `parse-capi` would it not then be fetched from the server? The nature of most-read is that it's regularly changing? If it were in the server side rendered response it'd be cached and out of date, or we'd have to bust the cache everytime it updates. Also we want the initial render to be as fast as possible, which means constructing the `CAPI` model as quickly as we can, waiting for `async` calls to respond will slow things down.

**Pull request merged**